nym-vpnd/src/service/error.rs

```diff
@@ -0,0 +1,123 @@
+use std::path::PathBuf;
+
+use nym_vpn_lib::credential_storage::error::StorageError;
+use nym_vpn_lib::credentials::CredentialError;
+use nym_vpn_lib::id::NymIdError;
+use time::OffsetDateTime;
+use tracing::error;
+
+#[derive(Clone, Debug, thiserror::Error)]
+pub enum ImportCredentialError {
+    #[error("vpn is connected")]
+    VpnRunning,
+
+    #[error("credential already imported")]
+    CredentialAlreadyImported,
+
+    #[error("generic error: {0}")]
+    Generic(String),
+
+    #[error("storage error: {0}")]
+    StorageError(String),
+
+    #[error("failed to deserialize credential: {reason}")]
+    DeserializationFailure { reason: String, location: PathBuf },
+
+    #[error("credential expired: {expiration}")]
+    CredentialExpired {
+        expiration: OffsetDateTime,
+        location: PathBuf,
+    },
+
+    // TODO: where is this coming from and is it really specific to freepass?
+    #[error("freepass expired: {expiration}")]
+    FreepassExpired { expiration: String },
+
+    #[error("failed to verify credential")]
+    VerificationFailed,
+
+    #[error("failed to query contract")]
+    FailedToQueryContract,
+}
+
+impl From<CredentialError> for ImportCredentialError {
+    fn from(err: CredentialError) -> Self {
+        let mut error = ImportCredentialError::Generic(err.to_string());
+        match err {
+            CredentialError::FailedToSetupStoragePaths { .. } => {}
+            CredentialError::FailedToCreateCredentialStoreDirectory { .. } => {}
+            CredentialError::FailedToReadCredentialStoreMetadata { .. } => {}
+            CredentialError::FailedToSetCredentialStorePermissions { .. } => {}
+            CredentialError::FailedToInitializePersistentStorage { .. } => {}
+            CredentialError::FreepassExpired { expiry_date } => {
+                error = ImportCredentialError::FreepassExpired {
```

Comment:
why do we mutate the `error` defined at the beginning rather than just returning here directly? all `mut` are spooky : )

Reply:
(then at the end, you could have a catch-all branch like `CredentialError::FailedToReadCredentialFile { .. } | CredentialError::FailedToGetNextUsableCredential { .. } | ... => ImportCredentialError::Generic(err.to_string())`

Reply:
having explicit `return` everywhere ended up being verbose, and I kinda wanted to list them all explicitly with the intention to incrementally start handling all of them.
However, I'm instead reworking the upstream error now to try to sort this out without this big nasty match